src/charm.py

```diff
@@ -1,104 +1,83 @@
-                    "override": "replace",
-                    "summary": "httpbin",
-                    "command": "gunicorn -b 0.0.0.0:80 httpbin:app -k gevent",
-                    "startup": "enabled",
-                    "environment": {"thing": self.model.config["thing"]},
-                }
-            },
-        }
-        # Add initial Pebble config layer using the Pebble API
-        container.add_layer("httpbin", pebble_layer, combine=True)
-        # Autostart any services that were defined with startup: enabled
-        container.autostart()
-        # Learn more about statuses in the SDK docs:
-        # https://juju.is/docs/sdk/constructs#heading--statuses
-        self.unit.status = ActiveStatus()
-
-    def _on_config_changed(self, _):
-        """Just an example to show how to deal with changed configuration.
-
-        TEMPLATE-TODO: change this example to suit your needs.
-        If you don't need to handle config, you can remove this method,
-        the hook created in __init__.py for it, the corresponding test,
-        and the config.py file.
-
-        Learn more about config at https://juju.is/docs/sdk/config
-        """
-        current = self.config["thing"]
-        if current not in self._stored.things:
-            logger.debug("found a new thing: %r", current)
-            self._stored.things.append(current)
-
-    def _on_fortune_action(self, event):
-        """Just an example to show how to receive actions.
-
-        TEMPLATE-TODO: change this example to suit your needs.
-        If you don't need to handle actions, you can remove this method,
-        the hook created in __init__.py for it, the corresponding test,
-        and the actions.py file.
-
-        Learn more about actions at https://juju.is/docs/sdk/actions
-        """
-        fail = event.params["fail"]
-        if fail:
-            event.fail(fail)
-        else:
-            event.set_results({"fortune": "A bug in the code is worth two in the documentation."})
+        self.name = "kafka"
+
+        self.framework.observe(getattr(self.on, "install"), self._on_install)
+        self.framework.observe(
+            getattr(self.on, "cluster_relation_joined"), self._on_cluster_relation_joined
+        )
+        self.framework.observe(getattr(self.on, "leader_elected"), self._on_leader_elected)
+        self.framework.observe(
+            getattr(self.on, "get_server_properties_action"), self._on_get_server_properties_action
+        )
+
+    @property
+    def _relation(self):
+        return self.model.get_relation("cluster")
+
+    def _on_install(self, _) -> None:
+        try:
+            self.unit.status = MaintenanceStatus("installing packages")
+            install_packages()
+            self.unit.status = ActiveStatus()
+        except (SnapError, PackageNotFoundError):
+            self.unit.status = BlockedStatus("failed to install packages")
+
+    def _on_leader_elected(self, _):
+        return
```

Comment:
is this hook needed? it doesnt do anything at the moment. the same for `_on_cluster_relation_joined`

Reply:
They're just stubs, will fill them in after my ZooKeeper PR gets resolved by ROCKs team.